rammingen_sdk

quasiyoke opened the pull request:
1. Detached common modules into a dedicated `rammingen_sdk` crate:
   1. `Cipher` — wrapper over AES-SIV-AEAD implementation.
   1. `Client` to `rammingen_server`. The client no longer tries to retry the request on application-level errors (it only retries on transport errors).
   1. `serve_connection` — a generic `hyper` error handler and graceful shutdown facilities.

   The GitHub diff shows that many modules were deleted and readded. In fact, most of them had their GIT history preserved, as you can see in the blame.

1. Several structures have been renamed:
   1. `ContentHandle` ← `DecryptedFileContent` — the word "content" here initially confused me and I assumed that the structure contains a blob of file contents.
   1. `EntryVersionHandle` ← `DecryptedEntryVersionData`

   I'd like to avoid using the word "decrypted" since this is the default state of data. I'd like to also avoid using the word "data" in structure identifiers since structures usually represent only data.

1. The archive client no longer tries to retry the request on application-level errors (it only retries on transport errors).

Riateche commented on protocol/src/credentials.rs:
```diff
@@ -0,0 +1,84 @@
+
+#[derive(Clone, AsRef, Derivative, Deserialize, Serialize)]
+#[derivative(Debug)]
+pub struct AccessToken(#[derivative(Debug = "ignore")] String);
```
I think in this case it's simpler to implement `Debug` manually than use `derivative`.

Riateche commented on protocol/src/credentials.rs:
```diff
@@ -0,0 +1,84 @@
+            "invalid length; got {}, expected {ACCESS_TOKEN_LENGTH}",
+            s.len(),
+        );
+        Ok(Self(s.to_owned()))
```
I think we should also restrict access token to be alphanumeric.

Riateche commented on protocol/src/credentials.rs:
```diff
@@ -0,0 +1,84 @@
+
+#[derive(Clone, Derivative)]
+#[derivative(Debug)]
+pub struct EncryptionKey(#[derivative(Debug = "ignore")] GenericArray<u8, U64>);
```
I think in this case it's simpler to implement `Debug` manually than use `derivative`.

Riateche commented on protocol/src/credentials.rs:
```diff
@@ -0,0 +1,84 @@
+
+    /// We intentionally don't implement `Display` on `Self` to prevent accidental
+    /// encryption key disclosure (in logs, etc).
+    pub fn fmt(&self) -> impl Display + '_ {
```
I think it would be nice to have a more verbose name, e.g. `display_unmasked`.

`AccessToken` probably should have a method like this too.

quasiyoke replied:
For `AccessToken`, instead of making a symmetric `fn display_unmasked(&self) → impl Display + '_`, I made `fn as_unmasked_str(&self) → &str`. `display_unmasked` would be ideal for specifying the token in HTTP headers (it accepts `impl Display`), but `impl sqlx::Encode` is needed to specify the token in SQL queries — it could be implemented, but it would force us to bring even more `sqlx` imports into the protocol.

Riateche commented on sdk/src/client/download.rs:
```diff
@@ -0,0 +1,102 @@
+    pub async fn download_and_decrypt(
+        &self,
+        handle: &ContentHandle,
+        path: impl AsRef<Path> + Display,
```
If it's just for logging, I would prefer to require `Debug` instead of `Display`.

Riateche commented on sdk/src/client/download.rs:
```diff
@@ -0,0 +1,102 @@
+            self.reqwest
+                .get(url)
+                .bearer_auth(self.token.as_ref())
+                .timeout(DEFAULT_TIMEOUT)
```
This has to be `3600 * 24`, not `DEFAULT_TIMEOUT`. Otherwise, download will fail for large files. The `timeout(DEFAULT_TIMEOUT, ...)` above is fine because it only applies to receiving the header, but `reqwest`'s timeout applies to the whole download.

Riateche commented on sdk/src/client/mod.rs:
```diff
@@ -0,0 +1,243 @@
+            Ok(x) => break Ok(x),
+            Err(RequestError::Application(err)) => break Err(err),
+            Err(RequestError::Transport(err)) if attempt >= NUM_RETRIES => break Err(err),
+            Err(RequestError::Transport(err)) => err,
```
I'm not a big fan of condition guards in `match` because it can be hard to reason about exact conditions for branch execution. I'd prefer this to be:
```suggestion
            Err(RequestError::Transport(err)) => {
                if attempt >= NUM_RETRIES {
                    break Err(err);
                } else {
                    err
                }
            },
```

Riateche commented on sdk/src/content.rs:
```diff
@@ -5,14 +5,12 @@ use rammingen_protocol::{
 
 #[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct DecryptedFileContent {
+pub struct ContentHandle {
```
I think it's nice to keep "decrypted"/"encrypted" in the name because otherwise it's easy to mix them up. Also I'm not sure "handle" is the best name. There is no actual resource handle in this type, so something like "ContentInfo" could be a better name.

Riateche commented on sdk/src/content.rs:
```diff
@@ -71,29 +69,29 @@ impl LocalEntryInfo {
 
 #[derive(Debug, Serialize, Deserialize)]
-pub struct DecryptedEntryVersionData {
+pub struct EntryVersionHandle {
```
Same concerns as for `ContentHandle` naming above.

Riateche commented on sdk/src/crypto/io.rs:
```diff
@@ -0,0 +1,246 @@
+    }
+}
+
+pub fn encrypt_file(path: impl AsRef<Path>, cipher: &Cipher) -> Result<EncryptedFileData> {
```
It makes sense to make this a method on `Cipher` as well.

Riateche reviewed (changes requested):
Nice, thank you. I've made a few comments.

quasiyoke commented:
Returned the "decrypted" identifiers back. Removed the words "Data" and "Info" from the identifiers. Added the word "Head" in the "header" meaning next to the "Content" (if it's not the content itself).

Riateche reviewed (approved):
Thanks!
